R/query_all_versions.R

```diff
@@ -0,0 +1,114 @@
+    suppressMessages(suppressWarnings(adc_solr<- adc_solr %>%
+                                          stringr::str_extract_all("name>.*<") %>%
+                                          unlist() %>%
+                                          stringr::str_replace_all("name>|<", "")))
+    return(adc_solr)
+}
+
+
+#' Create data frame with Solr fields
+#'
+#' This is a helper function for 'query_all_versions' function. It simplifies solr queries
+#' and creates a one row data frame with the specified Solr fields as the columns.
+#'
+#' @author Sharis Ochs
+#'
+#' @param node (character) Specify the node where the object should be searched for.
+#' @param object_pid (character) PID for the object that you want to return information about.
+#' @param fields (list) List of fields that you want returned in the data frame. Default
+#' returns all non NULL fields.
+#'
+#' @return (dataframe)
+#'
+
+query_solr_metadata<- function(node, object_pid, fields = "all"){
+
+    # Case 1 ========================
+    # Case: if no fields are specified
+    suppressWarnings(if (fields == "all"){
+        # Query all fields
+        q <- paste0("documents:\"", object_pid, "\"")
+        df_query <- dataone::query(node, list(q=q,
+                                              fl= "*",
+                                              rows="5"),
+                                   as = "data.frame")
+    })
+
+    # Case 2 ========================
+    # Else: Fields are specified
+    suppressWarnings(if (fields != "all"){
+        q <- paste0("documents:\"", object_pid, "\"")
+        fl <- paste(fields, collapse=", ")
+        df_query <- dataone::query(node, list(q=q,
+                                              fl= fl,
+                                              rows="5"),
+                                   as = "data.frame")
+    })
+    return(df_query)
+}
+
+#' Solr query all versions of a PID
+#'
+#' This function uses a combination of get_all_versions and solr query to return the query fields
+#' for all versions of the specified PID. Each row of the resulting data frame corresponds to a version
+#' and the columns are the query fields.
+#'
+#' @author Sharis Ochs
+#'
+#' @param node (character) Specify the node where the object should be searched for.
+#' @param object_pid (character) PID for the object that you want to return information about.
+#' @param fields (list) List of fields that you want returned in the data frame. Default
+#' returns all non NULL fields.
+#'
+#' @return (dataframe)
+#'
+query_all_versions<- function(node, object_pid, fields = "all"){
+
+    ## Checks =========================
+    # Check that object exist
+    if (!(arcticdatautils::object_exists(node, object_pid))) {
+        stop('Object does not exist on specified node ')
+    }
+
+    # Get all solr fields
+    adc_solr<- get_solr_fields()
+
+    # Check that all specified fields are valid
+    suppressWarnings(if (fields != "all"){
```

Comment:
You can re-write this a bit to eliminate the for loop.  The trick is to use [] (indexing) if possible.  
Chapter 3 in R Inferno taught me how to do this.  

```
x <- c(1,2,71,10)
y <- seq(1:10)

# these two lines aren't necessary 
x %in% y    # produces T/F vector 
which(x %in% y)   # returns indices 

indices <- which(x %in% y)
message(x[-indices], " are not valid fields")
```